tests/src/rule-tester.js

```diff
@@ -0,0 +1,62 @@
+import { RuleTester } from 'eslint';
+import { version as eslintVersion } from 'eslint/package.json';
+import semver from 'semver';
+
+export const usingFlatConfig = semver.major(eslintVersion) >= 9;
+
+export function withoutAutofixOutput(test) {
+  return { ...test, output: usingFlatConfig ? null : test.code };
+}
+
+export class FlatCompatRuleTester {
+  constructor(testerConfig) {
+    this._tester = new RuleTester(FlatCompatRuleTester._flatCompat(testerConfig));
+  }
+
+  run(
+    ruleName,
+    rule,
+    tests,
+  ) {
+    this._tester.run(ruleName, rule, {
+      valid: tests.valid.map((t) => FlatCompatRuleTester._flatCompat(t)),
+      invalid: tests.invalid.map((t) => FlatCompatRuleTester._flatCompat(t)),
+    });
+  }
+
+  static _flatCompat(config) {
+    if (!config || !usingFlatConfig || typeof config === 'string') {
+      return config;
+    }
+
+    const obj = {
+      languageOptions: { parserOptions: {} },
+    };
+
+    for (const [key, value] of Object.entries(config)) {
+      if (key === 'parser') {
+        obj.languageOptions.parser = typeof value === 'string' ? require(value) : value;
+
+        continue;
+      }
+
+      if (key === 'parserOptions') {
+        for (const [option, val] of Object.entries(value)) {
+          if (option === 'ecmaVersion' || option === 'sourceType') {
+            obj.languageOptions[option] = val;
+
+            continue;
+          }
+
+          obj.languageOptions.parserOptions[option] = val;
+        }
+
+        continue;
+      }
+
+      obj[key] = value;
+    }
```

Comment:
could this be replaced by something a bit more explicit, using object spreads instead of a loop over entries?

Reply:
possibly, though it'd mean having to explicitly name every potential property some of which are not currently being used and might not be obvious if they're forgotten about (i.e. tests might happen to pass without the particular config).

Also noting this is the same logic I've used in other plugins, so using that implementation here would make it easier to apply changes if it was later discovered something was missing

Reply:
i meant something like, taking the known keys, and using `...rest` to grab the unknown ones and spreading those in, but i guess it's fine to keep as-is too.

Reply:
This may need a little tweaking but something to this effect 

```js
  const { parser, parserOptions = {}, ...remainingConfig  } = config;
  const { languageOptions } = remainingConfig;
  const { ecmaVersion, sourceType, ...remainingParserOptions } = parserOptions;
  const parserObj = typeof parser === 'string' ? require(parser) : parser;
  
  const obj = {
       languageOptions: {
         ...languageOptions,
         ...(parserObj ? { parser: parserObj } : {}),
         parserOptions: {
           ...(ecmaVersion ? { ecmaVersion } : {}),
           ...(sourceType? { sourceType } : {}),
           ...remainingParserOptions,
         },
       },
       ...remainingConfig,
  };
```

Reply:
except instead of `...(condition ? obj : {})` you can just do `...condition && obj`, yes, that looks good

Reply:
can this one still be addressed?